lisa/sut_orchestrator/azure/platform_.py

```diff
@@ -1921,19 +1934,51 @@ def _get_deployable_vhd_path(
                     if self._is_stuck_copying(blob_client, log):
                         # Delete the stuck vhd.
                         blob_client.delete_blob(delete_snapshots="include")
-                    elif original_key == cached_key:
-                        # If it exists, return the link, not to copy again.
-                        log.debug("the sas url is copied already, use it directly.")
-                        vhd_exists = True
+                    elif original_key and cached_key:
+                        if original_key == cached_key:
+                            log.debug("the sas url is copied already, use it directly.")
+                            vhd_exists = True
+                        else:
+                            log.debug("found cached vhd, but the hash key mismatched.")
                     else:
-                        log.debug("found cached vhd, but the hash key mismatched.")
+                        log.debug(
+                            "No md5 content either in original blob or current blob. "
+                            "Then no need to check the hash key"
+                        )
+                        vhd_exists = True
 
             if not vhd_exists:
-                blob_client.start_copy_from_url(
-                    original_vhd_path, metadata=None, incremental_copy=False
-                )
+                azcopy_path = self._azure_runbook.azcopy_path
+                log.debug(f"AzCopy path:{azcopy_path}")
+                if azcopy_path and os.path.exists(azcopy_path):
+                    sas_token = generate_sas_token(
+                        credential=self.credential,
+                        subscription_id=self.subscription_id,
+                        account_name=storage_name,
+                        resource_group_name=self._azure_runbook.shared_resource_group_name,  # noqa: E501
+                        write=True,
+                    )
+                    dst_vhd_sas_url = f"{full_vhd_path}?{sas_token}"
+                    log.info(f"copying vhd by azcopy {dst_vhd_name}")
+                    try:
+                        copy_blob_by_azcopy(
+                            azcopy_path, src_vhd_sas_url, dst_vhd_sas_url, log
+                        )
+                    except Exception as identifier:
+                        blob_client.delete_blob(delete_snapshots="include")
+                        raise LisaException(f"{identifier}")
+
+                    # Set metadata to mark the blob copied by AzCopy successfully
+                    metadata = {"AzCopyStatus": "Success"}
+                    blob_client.set_blob_metadata(metadata)
+                    log.info("vhd copied")
+                    return full_vhd_path
```

Comment:
Why return here, not at the end?

Reply:
The copy operation with AzCopy is synchronous. We don't need to call wait_copy_blob. So return here.
wait_copy_blob needs to be called when there is another copy operation in parallel.

Reply:
It still can reach the wait_copy_blob to make the logic consistent, and do an extra validation.

Reply:
Done